datafusion/catalog/src/async.rs

```diff
@@ -0,0 +1,765 @@
+    async_provider: Arc<dyn AsyncCatalogProvider>,
+}
+impl ResolvedCatalogProviderBuilder {
+    fn new(async_provider: Arc<dyn AsyncCatalogProvider>) -> Self {
+        Self {
+            cached_schemas: HashMap::new(),
+            async_provider,
+        }
+    }
+    fn finish(self) -> Arc<dyn CatalogProvider> {
+        let cached_schemas = self
+            .cached_schemas
+            .into_iter()
+            .filter_map(|(key, maybe_value)| {
+                maybe_value.map(|value| (key, value.finish()))
+            })
+            .collect();
+        Arc::new(ResolvedCatalogProvider { cached_schemas })
+    }
+}
+
+/// A catalog provider list that looks up catalogs in a cache
+///
+/// Instances are created by the [`AsyncCatalogProviderList::resolve`] method
+#[derive(Debug)]
+struct ResolvedCatalogProviderList {
+    cached_catalogs: HashMap<String, Arc<dyn CatalogProvider>>,
+}
+impl CatalogProviderList for ResolvedCatalogProviderList {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn register_catalog(
+        &self,
+        _name: String,
+        _catalog: Arc<dyn CatalogProvider>,
+    ) -> Option<Arc<dyn CatalogProvider>> {
+        unimplemented!("resolved providers cannot handle registration APIs")
+    }
+
+    fn catalog_names(&self) -> Vec<String> {
+        self.cached_catalogs.keys().cloned().collect()
+    }
+
+    fn catalog(&self, name: &str) -> Option<Arc<dyn CatalogProvider>> {
+        self.cached_catalogs.get(name).cloned()
+    }
+}
+
+/// A trait for schema providers that must resolve tables asynchronously
+///
+/// The [`SchemaProvider::table`] method _is_ asynchronous.  However, this is primarily for convenience and
+/// it is not a good idea for this method to be slow as this will cause poor planning performance.
+///
+/// It is a better idea to resolve the tables once and cache them in memory for the duration of
+/// planning.  This trait helps implement that pattern.
+///
+/// After implementing this trait you can call the [`AsyncSchemaProvider::resolve`] method to get an
+/// `Arc<dyn SchemaProvider>` that contains a cached copy of the referenced tables.  The `resolve`
+/// method can be slow and asynchronous as it is only called once, before planning.
+#[async_trait]
+pub trait AsyncSchemaProvider: Send + Sync {
+    /// Return the name of the schema provided by this provider
+    ///
+    /// If a table reference's schema name does not match this name then the reference will be ignored
+    /// when calculating the cached set of tables (this allows other providers to supply the table)
+    fn name(&self) -> &str;
+    /// Return the name of the catalog this provider belongs to
+    ///
+    /// If a table reference's catalog name does not match this name then the reference will be ignored
+    /// when calculating the cached set of tables (this allows other providers to supply the table)
+    fn catalog_name(&self) -> &str;
```

Comment:
When trying this API out I didn't fully understand this API (or what i should return) -- maybe if it is optional / has an advanced usecase we could provide a default implementation

```suggestion
    fn catalog_name(&self) -> Option<&str> { return None } 
```

Reply:
Yeah, I was a bit torn on this anyways.  The problem arises when the user wants to use an AsyncSchemaProvider or AsyncCatalogProvider as the top level catalog.  In these cases it isn't clear what we should do with full / partial table references.

For example, if the user adds an AsyncSchemaProvider and then tries to resolve the query "SELECT * FROM weston.public.my_table" what should we do?

 * We could just assume that all table references are intended for us.  This works as long as this schema provider is the only provider registered.  If there are multiple providers registered then we need to know which to use for a given table reference somehow.
 * We could assume we don't match the table reference and we will only match bare references.
 * Or we can require schema providers to supply their own name and the catalog name so that we can filter references that apply (this is what I do)

The main problem with the current approach is that users whose top level is an AsyncCatalogProviderList have to implement these methods even though they are meaningless (we will do the filtering in the higher level resolve function).

We should probably do whatever the synchronous planner does in this case but I just didn't know.  If I register a schema provider with a `SessionContext` and then call `sql` with a full (not bare) table reference then does it apply the provider I registered or not?